feat: CKF eta-dependent cuts

timadye opened the pull request:
Adds the ability to specify |eta| bins for cuts in the CKF algorithm.
 * add bin boundaries and change cuts to vectors in `Acts::MeasurementSelectorCuts`
 * Add `--ckf-selection-abseta-bins` option to CKF example
 * `--ckf-selection-chi2max` and `--ckf-selection-nmax` can now take different values for each eta bin, eg.

```
ActsExampleCKFTracksTGeo \
    --ckf-selection-abseta-bins=1:2.4:3.2 \
    --ckf-selection-chi2max=15:40:15:40 \
    --ckf-selection-nmax=10:15:10:10
```

timadye commented:
The ubuntu2004 test failed. It looks like there is the new truth tracking python example needs to be updated too.
I'm testing the Python bindings, so I'll be able to try this myself and add a fix to this PR.

timadye commented:
I fixed the pytest example and all checks now run OK.
@paulgessinger, can you remove the WIP tag and take a look?
Thanks, Tim.

paulgessinger reviewed (commented):
I have a few more or less minor comments.

The suggestions on `if() {}` are a bit on the stylistic side, I suppose.

timadye commented on Core/include/Acts/TrackFinding/MeasurementSelector.hpp:
```diff
@@ -26,10 +26,12 @@ namespace Acts {
+  std::vector<double> chi2CutOff{std::numeric_limits<double>::max()};
   /// Maximum number of associated measurements on a single surface.
-  size_t numMeasurementsCutOff = 1;
+  std::vector<int> numMeasurementsCutOff{1};
```
This was the simplest option so I could define the `--ckf-selection-nmax` option using the `VariableIntegers`. I don't thing there is `VariableUnsignedIntegers` type. With that choice, it was just a question where to do the conversion. I did it [here](https://github.com/acts-project/acts/pull/1064/files/a88294b7494069a2dac35467bce11b36739ce927#diff-a767add0929eb1306bbca2878f5ad6a9acfcce2ef465810001a6a37973e27292R152).
I could also move it somewhere else, though probably needing a vector conversion.

paulgessinger replied:
I since this is disconnected from the CLI inputs, I guess I would lean towards using `unsigned int` here and converting the CLI arguments before they hit `MeasurementSelector`. I can't think of any use-case where we would want to have negative number of measurements 😆

timadye replied:
OK, I fixed that.
It will still give a crazy-big cut if you do specify a negative number, but that's now just in the examples no longer in the Core.
Thanks for the suggestion.

timadye commented on Core/include/Acts/TrackFinding/MeasurementSelector.hpp:
```diff
@@ -184,6 +187,28 @@ class MeasurementSelector {
 
  private:
+  template <typename CutType>
+  static CutType VariableCut(
```
The idea is that this routine could also add other variable cuts, eg. in pT bins. That's why it is passed `selector`, rather than `selector->etaBins`. Such a change would still need changes to `MeasurementSelector::Config`, but not any of its uses. This might be a moot point if the same routine can be used elsewhere and I move it to a Utility function.
It might be clearer if this were a static member of `MeasurementSelector::Config`, but I think it's clearer if we keep that unambiguously POD (if that's the right term for something with non-POD members).

paulgessinger replied:
Ah right, so you mean that even though it specifically looks into `selector->etaBins` right now, it might be extended to also look into other members of `selector`?

timadye replied:
Yes, exactly :)

paulgessinger replied:
Ok.

timadye commented:
> The suggestions on `if() {}` are a bit on the stylistic side, I suppose.

Yeah, I actually prefer the single-line style (`if (test) action;`) but the acts clang-format style turned them into 2 lines. So 3 lines is just more of the same.

If this is a recommended format, can it be added to the clang-format? Or isn't that technically possible?

paulgessinger commented:
Unfortunately, as far as I know `clang-format` can't detect this. `clang-tidy` can, but this currently doesn't run in the CI because it's notoriously brittle and takes very long.

timadye commented:
This PR was incompatible with #1062's `MeasurementSelector`, so I rebased, squashed some minor updates, moved my changes from `MeasurementSelector.hpp` to `MeasurementSelector.cpp`, and changed the logic a bit to make it more efficient. This all compiles OK, so I force-pushed the new version.

Unfortunately, it doesn't seem to get the correct value of
```eta = Acts::VectorHelpers::eta(trackState.predicted())```
which worked before using
```eta = Acts::VectorHelpers::eta(predictedParams.unitDirection())```

I've marked this PR as WIP until I can fix it.

paulgessinger commented:
Ah right, sorry, I didn't realize this before. The track state proxy only has the local bound parameter values, and not the global ones, which are needed for calculating eta. 

In principle, the reference surface should be available though, so potentially that could be used to select the eta slice (maybe in combination with the local parameters)?

timadye commented:
Thanks for the hint. I'll try that.

It's a bit naughty that `Acts::VectorHelpers::eta()` et al seem to work (compile and run) but return the wrong value if you pass the wrong type of parameter value. I guess the problem is they are implemented as Eigen vectors with different interpretations of the elements. I guess that's needed to allow different types in the proxy. Or could you wrap the matrix in a concrete type that said these were local bound parameters?

paulgessinger commented:
`Acts::VectorHelpers::eta()` really only does `std::atanh(v[2] / v.norm())`, so interprets it as `xyz`, like what it got from `unitDirection`. `.predicted()` is a 6D vector, and as it stands `::eta()` only checks whether there are **at least** 3 components.

1. Do you think we (I) should adjust `eta()` to check for exactly 3 components (at compile time) for a bit of extra safety?
2. I guess you'd have to convert to an `xyz` vector before calling `eta()` in any case, since that function has no information on the parametrization at all.

timadye commented:
> Do you think we (I) should adjust `eta()` to check for exactly 3 components (at compile time) for a bit of extra safety?

This sounds like a good idea. I guess it could cause trouble if we ever use a parameterisation that has x,y,z as only the first three parameters, eg. x,y,z,px,py,pz or x,y,z,t. But if those aren't too common, I guess it would be possible to explicitly pass a subvector.

It wouldn't catch cases where you use a different 3-component parameterisation, eg. (R, theta, phi), but I guess we don't have many of them, at least with GPDs that use eta.

Would you also do this for the other similar functions `phi()` and `perp()`?

I was actually thinking of using concrete types (rather than raw Eigen matrix types) for all these things. But I guess that's a much (much!) bigger change, and probably already discussed to death. What you suggest sounds like an easy fix for this case.

paulgessinger reviewed (approved):
LGTM
